lightning/src/ln/async_payments_tests.rs

```diff
@@ -1272,3 +1280,754 @@ fn expired_static_invoice_payment_path() {
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+	create_unannounced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0);
+	let server = &nodes[0];
+	let recipient = &nodes[1];
+
+	let recipient_id = vec![42; 32];
+	let inv_server_paths =
+		server.node.blinded_paths_for_async_recipient(recipient_id, None).unwrap();
+	recipient.node.set_paths_to_static_invoice_server(inv_server_paths).unwrap();
+
+	// Up to TEST_MAX_UPDATE_ATTEMPTS offer_paths_requests are allowed to be sent out before the async
+	// recipient should give up.
+	for _ in 0..TEST_MAX_UPDATE_ATTEMPTS {
+		recipient.node.test_check_refresh_async_receive_offers();
+		let offer_paths_req = recipient
+			.onion_messenger
+			.next_onion_message_for_peer(server.node.get_our_node_id())
+			.unwrap();
+		assert!(matches!(
+			server.onion_messenger.peel_onion_message(&offer_paths_req).unwrap(),
+			PeeledOnion::AsyncPayments(AsyncPaymentsMessage::OfferPathsRequest(_), _, _)
+		));
+		recipient.onion_messenger.release_pending_msgs(); // Ignore redundant paths requests
+	}
+
+	// After the recipient runs out of attempts to request offer paths, they will give up until the
+	// next timer tick.
+	recipient.node.test_check_refresh_async_receive_offers();
+	assert!(recipient
+		.onion_messenger
+		.next_onion_message_for_peer(server.node.get_our_node_id())
+		.is_none());
+
+	// On the next timer tick, more offer paths requests should be allowed to go through.
+	recipient.node.timer_tick_occurred();
+	let offer_paths_req = recipient
+		.onion_messenger
+		.next_onion_message_for_peer(server.node.get_our_node_id())
+		.unwrap();
+	assert!(matches!(
+		server.onion_messenger.peel_onion_message(&offer_paths_req).unwrap(),
+		PeeledOnion::AsyncPayments(AsyncPaymentsMessage::OfferPathsRequest(_), _, _)
+	));
+}
+
+#[test]
+fn limit_serve_static_invoice_requests() {
+	// If we have enough async receive offers cached already, the recipient should stop sending out
+	// offer_paths_requests.
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let persister;
+	let chain_monitor;
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let payee_node_deserialized;
+	let mut nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+	let chan_id =
+		create_unannounced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0).0.channel_id;
+	let server = &nodes[0];
+	let recipient = &nodes[1];
+
+	let recipient_id = vec![42; 32];
+	let inv_server_paths =
+		server.node.blinded_paths_for_async_recipient(recipient_id.clone(), None).unwrap();
+	recipient.node.set_paths_to_static_invoice_server(inv_server_paths).unwrap();
+
+	// Build the target number of offers interactively with the static invoice server.
```

Comment:
I like it that you put these comments inline in the test to describe and keep apart the steps.